### PR TITLE
Add facility to convert tsc cycles to time

### PR DESCRIPTION
**Pull request by nsavoire**
# What does this PR do?

Allow conversion from rdtsc values to time.

### CODE REVIEW

**Comment by r1viollet** [comment]
Not what I was expecting when I opened the PR ! 😄  Thanks for taking this into account

File: include/timer.hpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+using TscCycles = uint64_t;
+
+#ifdef __x86_64__
+inline TscCycles read_tsc() { return __rdtsc(); }
+#elif defined(__aarch64__)
+inline TscCycles read_tsc() {
+  uint64_t val;
+
+  asm volatile("mrs %0, cntvct_el0" : "=r"(val));
```
[the remark this answers is not in the document]
Comment (nsavoire) [reply include/timer.hpp:35]: Most of this code is taken from dpdk (https://github.com/DPDK/dpdk/blob/main/lib/eal/arm/include/rte_cycles_64.h)
I believe this counter is comparable to TSC and does not require elevated privileges. Its frequency is in the range 0-100Mhz.
There is a higher resolution PMU counter, but access to it is not configured by default in the kernel.

File: src/timer.cc [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+static int init_from_perf(ddprof::TscConversion &conv) {
+  perf_event_attr pe = {.type = PERF_TYPE_SOFTWARE,
+                        .size = sizeof(struct perf_event_attr),
+                        .config = PERF_COUNT_SW_DUMMY,
+                        .disabled = 1,
+                        .exclude_kernel = 1,
+                        .exclude_hv = 1};
+  int fd = perf_event_open(&pe, 0, -1, -1, 0);
```
[the remark this answers is not in the document]
Comment (nsavoire) [reply src/timer.cc:157]: Thanks for the tip, I will change that !

File: src/timer.cc [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+
+#endif
+
+uint64_t get_tsc_freq() {
+  timespec sleeptime = {.tv_nsec = k_ns_per_sec / 10}; /* 1/10 second */
+
+  timespec t_start;
+  if (clock_gettime(CLOCK_MONOTONIC_RAW, &t_start) == 0) {
+    uint64_t start = ddprof::read_tsc();
+    nanosleep(&sleeptime, NULL);
```
[the remark this answers is not in the document]
Comment (nsavoire) [reply src/timer.cc:133]: The code does not check if constant and nonstop TSC is present. If we can get calibration information from perf, then it means that TSC is invariant, but there are cases where even if we can't get the info from perf, TSC still provides accurate measurements (eg. on my laptop).

File: src/timer.cc [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+  if (family == 6 || family == 15) {
+    uint32_t ext_model = (fam_mod_step >> 16) & 0xf;
+    model += (ext_model << 4);
+  }
+
+  return model;
+}
+
+static int32_t rdmsr(int msr, uint64_t *val) {
+  int fd = open("/dev/cpu/0/msr", O_RDONLY);
```
[the remark this answers is not in the document]
Comment (nsavoire) [reply src/timer.cc:41]: It's part of the fallback process from dpdk, I guess it does not hurt to have it.

**Review by sanchda — APPROVED** [review]
Very nice!  LGTM

**Pull request merged**